anta/tests/security.py

```diff
@@ -654,3 +655,151 @@ def test(self) -> None:
+            return
+
+        # Iterate over all ip sec connection
+        for connection, conn_data in command_output.items():
+            state = next(iter(conn_data["pathDict"].values()))
+            if state != "Established":
+                failure_conn.append(connection)
+        if failure_conn:
+            failure_msg = "\n".join(failure_conn)
+            self.result.is_failure(f"Following IPv4 security connections are not establised:\n{failure_msg}.")
+
+
+class VerifySpecificIPSecConn(AntaTest):
+    """
+    Verifies IPv4 security connections state for a peer.
+
+    Expected Results
+    ----------------
+    * Success: The test passes if the IPv4 security connection for a peer is established in the specified VRF.
+    * Failure: The test fails if IPv4 security is not configured, a connection is not found for a peer, or the connection is not established in the specified VRF.
+
+    Examples
+    --------
+    ```yaml
+    anta.tests.security:
+      - VerifySpecificIPSecConn:
+          ip_security_connections:
+            - peer: 10.255.0.1
+            - peer: 10.255.0.2
+            vrf: default
+            connection:
+              - source_address: 100.64.3.2
+                estination_address: 100.64.1.2
+              - source_address: 172.18.3.2
+                destination_address: 172.18.2.2
+    ```
+    """
+
+    name = "VerifySpecificIPSecConn"
+    description = "Verifies IPv4 security connections for a peer."
+    categories: ClassVar[list[str]] = ["security"]
+    commands: ClassVar[list[AntaCommand | AntaTemplate]] = [AntaTemplate(template="show ip security connection vrf {vrf} path peer {peer}")]
+
+    class Input(AntaTest.Input):
+        """Input model for the VerifySpecificIPSecConn test."""
+
+        ip_security_connections: list[IPSecPeers]
+        """List of IP4v security peers."""
+
+        class IPSecPeers(BaseModel):
+            """Details of IPv4 security peers."""
+
+            peer: IPv4Address
+            """IPv4 address of the peer."""
+
+            vrf: str = "default"
+            """This is the optional VRF for the IP security peer. It defaults to `default` if not provided."""
+
+            connection: list[IPSecConn] | None = None
+            """Optional list of IPv4 security connections of a peer."""
+
+            class IPSecConn(BaseModel):
+                """Details of IPv4 security connections for a peer."""
+
+                source_address: IPv4Address
+                """Source address of the connection."""
+                destination_address: IPv4Address
+                """Destination address of the connection."""
+
+    def render(self, template: AntaTemplate) -> list[AntaCommand]:
+        """Render the template for each input IP Sec connection."""
+        return [template.render(peer=conn.peer, vrf=conn.vrf, connections=conn.connection) for conn in self.inputs.ip_security_connections]
```

Comment:
so we want to stop overloading the template.params this way 
can you move the connection out and reparse the inputs in the test function?

Reply:
updated in ba7f234